File: lib/token.ml
type t =
  (* Keywords *)
  | Let
  | Open
  | Import
  | Module
  | Export
  | Match
  | With
  | As
  | Return
  | If
  | Else
  | Struct
  | Enum
  | Macro
  | Derive
  | Mut
  | Defer
  | True
  | False
  (* Tokens *)
  (* - Double *)
  | Double_colon (* :: *)
  | Fat_arrow (* => *)
  | Double_eql
  | Skinny_arrow (* -> *)
  | Back_arrow (* <- *)
  | Pipe (* |> *)
  | Or (* || *)
  | Bitwise_or (* |. *)
  | And (* && *)
  | Bitwise_and (* &. *)
  | Lesser_eql
  | Greater_eql
  | Ellipsis (* ... *)
  | Walrus (* := *)
  (* - Single*)
  | Tilde
  | Back_tick
  | Bang
  | At
  | Hash
  | Dollar
  | Mod
  | Carrot
  | Amper
  | Star
  | Open_paren
  | Close_paren
  | Open_brack
  | Close_brack
  | Open_square
  | Close_square
  | Low_dash
  | Dash
  | Plus
  | Eql
  | Bar
  | Back_slash
  | Colon
  | Semi_colon
  | Lesser
  | Comma
  | Greater
  | Dot
  | Question
  | Forward_slash
  (* Literals *)
  | String of string
  | Char of char
  | Ident of string
  | Number of int
  | Eof
  | Unknown of char
[@@deriving show]

type loc =
  { file : string
  ; row : int
  ; col : int
  }

let show_loc { file; row; col } = Printf.sprintf "%s:%d:%d" file row col

let from_string = function
  | "let" -> Let
  | "open" -> Open
  | "import" -> Import
  | "module" -> Module
  | "export" -> Export
  | "match" -> Match
  | "as" -> As
  | "with" -> With
  | "return" -> Return
  | "if" -> If
  | "else" -> Else
  | "struct" -> Struct
  | "enum" -> Enum
  | "macro" -> Macro
  | "derive" -> Derive
  | "mut" -> Mut
  | "defer" -> Defer
  | "true" -> True
  | "false" -> False
  | s -> Ident s
;;

File: lexer/src/token.rs
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Span {
    pub file: String,
    pub line: usize,
    pub column: usize,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TokenKind {
    // Keywords
    Let,    // let
    Open,   // open
    Import, // import
    Module, // module
    Export, // export
    Match,  // match
    With,   // with
    As,     // as
    Return, // return
    If,     // if
    While,  // while
    For,    // for
    In,     // in
    Else,   // else
    Struct, // struct
    Enum,   // enum
    Derive, // derive
    Mut,    // mut
    Defer,  // defer
    True,   // true
    False,  // false
    // Double character tokens
    DoubleColon, // ::
    FatArrow,    // =>
    DoubleEql,   // ==
    NotEql,      // !=
    ShiftLeft,   // <<
    ShiftRight,  // >>
    PlusPlus,    // ++
    MinusMinus,  // --
    SkinnyArrow, // ->
    BackArrow,   // <-
    Pipe,        // |>
    Or,          // ||
    BitwiseOr,   // |.
    And,         // &&
    BitwiseAnd,  // &.
    LesserEql,   // <=
    GreaterEql,  // >=
    Ellipsis,    // ...
    Walrus,      // :=
    Concat,      // ~
    PlusEql,     // +=
    MinusEql,    // -=
    StarEql,     // *=
    SlashEql,    // /=
    // Single character tokens
    Tilde,              // ~
    BackTick,           // `
    Bang,               // !
    At,                 // @
    Hash,               // #
    Dollar,             // $
    Mod,                // %
    Carrot,             // ^
    Amper,              // &
    Star,               // *
    OpenParen,          // (
    CloseParen,         // )
    OpenBrack,          // {
    CloseBrack,         // }
    OpenSquare,         // [
    CloseSquare,        // ]
    LowDash,            // _
    Dash,               // -
    Plus,               // +
    Eql,                // =
    Bar,                // |
    BackSlash,          // \
    Colon,              // :
    SemiColon,          // ;
    Lesser,             // <
    Comma,              // ,
    Greater,            // >
    Dot,                // .
    Question,           // ?
    ForwardSlash,       // /
    String(String),     // "string"
    Char(char),         // 'c'
    DocComment(String), // (*! comment *)
    Ident(String),      // identifier
    Number(i32),        // 1234
    Unknown(char),      // unknown character
    Eof,                // end of file
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokenKind,
    pub span: Span,
}
